File: matlab/src/api/cache/cache_erec.m
%{ @file cache_erec.m
 %  @brief Recursive computation of the normalizing constant for cache models
 %
 %  @author LINE Development Team
%}

%{
 % @brief Computes the normalizing constant for cache models recursively
 %
 % @details
 % This function recursively computes the normalizing constant E for
 % cache models with given item popularity probabilities and cache capacity.
 % With item sizes and per-list storage cost caps it evaluates instead the
 % constrained normalizing constant E(m,k) of Casale-Gast (IEEE/ACM ToN,
 % 2021), Sec. IX, using the recursion
 %   E(m,k) = E_i(m,k) + sum_j m_j gamma_ij E_i(m-1_j, k-sigma_i 1_j)
 % with the extra boundary E(m,k)=0 whenever some k_j is negative.
 %
 % @par Syntax:
 % @code
 % E = cache_erec(gamma, m)
 % E = cache_erec(gamma, m, sigma, k)
 % @endcode
 %
 % @par Parameters:
 % <table>
 % <tr><th>Name<th>Description
 % <tr><td>gamma<td>Item popularity probabilities
 % <tr><td>m<td>Cache capacity
 % <tr><td>sigma<td>(Optional) item storage costs (sizes), 1 x n, positive integers
 % <tr><td>k<td>(Optional) per-list storage cost caps, 1 x h, non-negative integers
 % </table>
 %
 % @par Returns:
 % <table>
 % <tr><th>Name<th>Description
 % <tr><td>E<td>Normalizing constant
 % </table>
%}
function E=cache_erec(gamma,m,sigma,k)
if nargin<3 || isempty(sigma) || isempty(k)
    E = sub_cache_erec(gamma,m,length(gamma));
else
    E = sub_cache_erec_cost(gamma,m,sigma(:).',k(:).');
end
end

function E=sub_cache_erec(gamma,m,k)
h=length(m);
if sum(m)==0
    E=1;
    return
end
if sum(m)>k || min(m)<0
    E=0;
    return
end

if k==1 && sum(m)==1
    j = find(m);
    E=gamma(1,j);
    return
end
E = sub_cache_erec(gamma,m,k-1);
for j=1:h
    if m(j)>0
        E = E + gamma(k,j)*m(j)*sub_cache_erec(gamma,oner(m,j),k-1);
    end
end
end

% Dynamic program over the (residual capacity, residual cost cap) lattice.
function E=sub_cache_erec_cost(gamma,m,sigma,k)
[n,h] = size(gamma);
m = m(:).';
if numel(sigma)~=n
    line_error(mfilename,'The item size vector must have one entry per item.');
end
if numel(k)~=h
    line_error(mfilename,'The cost cap vector must have one entry per cache list.');
end
if any(sigma<=0) || any(abs(sigma-round(sigma))>0)
    line_error(mfilename,'Item sizes must be positive integers.');
end
if any(k<0) || any(abs(k-round(k))>0)
    line_error(mfilename,'Storage cost caps must be non-negative integers.');
end
if sum(m)>n || min(m)<0
    E=0;
    return
end
if sum(m)==0
    E=1;
    return
end
dims = [m+1, k+1];
lattice = prod(dims);
if lattice > 1e7
    line_error(mfilename,'The cost-constrained normalizing constant lattice has %d states, which exceeds the exact method limit; use the sampling method.', lattice);
end
% strides for a column-major linear index over dims
stride = cumprod([1, dims(1:end-1)]);
% F(mm,kk) over items 1..t; at t=0 only the empty cache contributes
F = zeros(lattice,1);
mcount = zeros(lattice,1);  % sum(mm) for each lattice point
sub = zeros(1,2*h);
for idx = 1:lattice
    rem = idx-1;
    for d = 2*h:-1:1
        sub(d) = floor(rem/stride(d));
        rem = rem - sub(d)*stride(d);
    end
    mcount(idx) = sum(sub(1:h));
end
F(mcount==0) = 1;
for t = 1:n
    Fprev = F;
    for idx = 1:lattice
        if mcount(idx)>t
            F(idx) = 0;
            continue
        end
        val = Fprev(idx);
        rem = idx-1;
        for d = 2*h:-1:1
            sub(d) = floor(rem/stride(d));
            rem = rem - sub(d)*stride(d);
        end
        for j = 1:h
            mj = sub(j);
            kj = sub(h+j);
            if mj>0 && kj>=sigma(t) && gamma(t,j)~=0
                val = val + gamma(t,j)*mj*Fprev(idx - stride(j) - sigma(t)*stride(h+j));
            end
        end
        F(idx) = val;
    end
end
E = F(end);
end

File: jar/src/main/java/jline/api/cache/Cache_erec.java
/**
 * @file Exact Recursive Cache Analysis
 *
 * Implements exact recursive (EREC) algorithms for cache system analysis.
 * Provides numerically exact solutions for cache performance metrics in
 * systems where computational complexity allows for precise evaluation.
 *
 * @since LINE 3.0
 */
package jline.api.cache;

import jline.io.InputOutput;
import jline.util.matrix.Matrix;

public final class Cache_erec {
    private Cache_erec() {}

    /**
     * Computes the cache miss rate using an exact recursive method.
     *
     * @param gamma Matrix representing the cache access factors.
     * @param m     Matrix representing the cache capacity vector.
     * @return Matrix containing the computed cache miss rates.
     */
    public static Matrix cache_erec(Matrix gamma, Matrix m) {
        return cache_erec_aux(gamma, m, gamma.getNumRows());
    }

    /**
     * Computes the normalizing constant under per-list storage cost caps,
     * E(m,k) = E_i(m,k) + sum_j m_j gamma_ij E_i(m-1_j, k-sigma_i 1_j), with the
     * extra boundary E(m,k)=0 whenever some residual cap is negative. See
     * Casale-Gast, IEEE/ACM Trans. Networking 29(2), 2021, Sec. IX.
     *
     * @param gamma Matrix representing the cache access factors.
     * @param m     Matrix representing the cache capacity vector.
     * @param sigma Item storage costs (sizes), one per item; null or empty for none.
     * @param k     Per-list storage cost caps, one per list; null or empty for none.
     * @return Matrix containing the computed normalizing constant.
     */
    public static Matrix cache_erec(Matrix gamma, Matrix m, Matrix sigma, Matrix k) {
        if (sigma == null || k == null || sigma.isEmpty() || k.isEmpty()) {
            return cache_erec(gamma, m);
        }
        return Matrix.singleton(cache_erec_cost(gamma, m, sigma, k));
    }

    /**
     * Dynamic program over the (residual capacity, residual cost cap) lattice.
     */
    private static double cache_erec_cost(Matrix gamma, Matrix m, Matrix sigma, Matrix k) {
        int n = gamma.getNumRows();
        int h = m.length();
        if (sigma.length() != n) {
            InputOutput.line_error(InputOutput.mfilename(new Object()),
                    "The item size vector must have one entry per item.");
        }
        if (k.length() != h) {
            InputOutput.line_error(InputOutput.mfilename(new Object()),
                    "The cost cap vector must have one entry per cache list.");
        }
        int[] mv = new int[h];
        int[] kv = new int[h];
        for (int j = 0; j < h; j++) {
            mv[j] = (int) Math.round(m.get(j));
            kv[j] = (int) Math.round(k.get(j));
            if (mv[j] < 0) {
                return 0.0;
            }
            if (kv[j] < 0) {
                return 0.0;
            }
        }
        int[] sv = new int[n];
        int mtot = 0;
        for (int j = 0; j < h; j++) {
            mtot += mv[j];
        }
        for (int i = 0; i < n; i++) {
            sv[i] = (int) Math.round(sigma.get(i));
            if (sv[i] <= 0) {
                InputOutput.line_error(InputOutput.mfilename(new Object()),
                        "Item sizes must be positive integers.");
            }
        }
        if (mtot > n) {
            return 0.0;
        }
        if (mtot == 0) {
            return 1.0;
        }
        int[] dims = new int[2 * h];
        for (int j = 0; j < h; j++) {
            dims[j] = mv[j] + 1;
            dims[h + j] = kv[j] + 1;
        }
        long lattice = 1;
        for (int d = 0; d < 2 * h; d++) {
            lattice *= dims[d];
            if (lattice > 10000000L) {
                InputOutput.line_error(InputOutput.mfilename(new Object()),
                        "The cost-constrained normalizing constant lattice exceeds the exact method limit; use the sampling method.");
            }
        }
        int size = (int) lattice;
        int[] stride = new int[2 * h];
        stride[0] = 1;
        for (int d = 1; d < 2 * h; d++) {
            stride[d] = stride[d - 1] * dims[d - 1];
        }
        int[] sub = new int[2 * h];
        double[] F = new double[size];
        double[] Fprev = new double[size];
        for (int idx = 0; idx < size; idx++) {
            decode(idx, stride, sub);
            int c = 0;
            for (int j = 0; j < h; j++) {
                c += sub[j];
            }
            F[idx] = (c == 0) ? 1.0 : 0.0;
        }
        for (int t = 1; t <= n; t++) {
            System.arraycopy(F, 0, Fprev, 0, size);
            for (int idx = 0; idx < size; idx++) {
                decode(idx, stride, sub);
                int c = 0;
                for (int j = 0; j < h; j++) {
                    c += sub[j];
                }
                if (c > t) {
                    F[idx] = 0.0;
                    continue;
                }
                double val = Fprev[idx];
                for (int j = 0; j < h; j++) {
                    int mj = sub[j];
                    int kj = sub[h + j];
                    double g = gamma.get(t - 1, j);
                    if (mj > 0 && kj >= sv[t - 1] && g != 0.0) {
                        val += g * mj * Fprev[idx - stride[j] - sv[t - 1] * stride[h + j]];
                    }
                }
                F[idx] = val;
            }
        }
        return F[size - 1];
    }

    private static void decode(int idx, int[] stride, int[] sub) {
        int rem = idx;
        for (int d = stride.length - 1; d >= 0; d--) {
            sub[d] = rem / stride[d];
            rem -= sub[d] * stride[d];
        }
    }

    /**
     * Auxiliary method for computing the cache miss rate using an exact recursive method.
     *
     * @param gamma Matrix representing the cache access factors.
     * @param m     Matrix representing the cache capacity vector.
     * @param k     Integer representing the current number of rows in the recursive step.
     * @return Matrix containing the computed cache miss rates.
     */
    public static Matrix cache_erec_aux(Matrix gamma, Matrix m, int k) {
        int h = m.getNumElements();

        if (m.elementSum() == 0.0) {
            return Matrix.singleton(1.0);
        }

        if (m.elementSum() > k || m.elementMin() < 0) {
            return Matrix.singleton(0.0);
        }

        if (k == 1 && m.elementSum() == 1.0) {
            int j = (int) m.find().value(); // Find the index of the non-zero element in m
            return Matrix.singleton(gamma.get(0, j));
        }

        Matrix E = cache_erec_aux(gamma, m, k - 1);
        for (int j = 0; j < h; j++) {
            if (m.get(j) > 0) {
                Matrix onerM = Matrix.oner(m, j);
                Matrix term = cache_erec_aux(gamma, onerM, k - 1).scale(gamma.get(k - 1, j) * m.get(j));
                E = E.add(term);
            }
        }
        return E;
    }
}
